### PR TITLE
Issue #99 : Dispose ILoggerProvider 

### PR DESCRIPTION
**Pull request by ycrumeyrolle**
- LoggerFactory disposes its providers
 - TraceSourceLoggerProvider flush & dispose its root TraceListener

### CODE REVIEW

**Comment by msotclas** [comment]
Hi __@ycrumeyrolle__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft Open Technologies, Inc. and real humans are currently evaluating your PR.</p>

TTYL, MSOTBOT;

File: src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs [hunk 1]
```diff
@@ -11,13 +11,15 @@ namespace Microsoft.Framework.Logging
     /// <summary>
     /// Provides an ILoggerFactory based on System.Diagnostics.TraceSource.
     /// </summary>
-    public class TraceSourceLoggerProvider : ILoggerProvider
+    public class TraceSourceLoggerProvider : ILoggerProvider, IDisposable
```
Comment (lodejard) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: Should ILoggerProvider inherit IDisposable?

Comment (davidfowl) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: https://github.com/aspnet/Logging/issues/99

Comment (ycrumeyrolle) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: So, what should be done ?

Comment (kichalla) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: @davidfowl Any update on this? Should @ycrumeyrolle follow @lodejard 's suggestion?

File: src/Microsoft.Framework.Logging/LoggerFactory.cs [hunk 2]
```diff
@@ -10,11 +10,12 @@ namespace Microsoft.Framework.Logging
     /// <summary>
     /// Summary description for LoggerFactory
     /// </summary>
-    public class LoggerFactory : ILoggerFactory
+    public class LoggerFactory : ILoggerFactory, IDisposable
```
Comment (kichalla) [inline src/Microsoft.Framework.Logging/LoggerFactory.cs]: The dispose on this factory should be called in Hosting...do you want us to take care of it?
https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNet.Hosting/WebHostBuilder.cs#L53

File: src/Microsoft.Framework.Logging/LoggerFactory.cs [hunk 3]
```diff
@@ -48,5 +49,31 @@ internal ILoggerProvider[] GetProviders()
...
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                    foreach (var provider in _providers)
+                    {
+                        var disposable = provider as IDisposable;
+                        if (disposable!= null)
+                        {
+                            disposable.Dispose();
```
Comment (kichalla) [inline src/Microsoft.Framework.Logging/LoggerFactory.cs]: Ideally Dispose shouldn't throw but I would put a try-catch around this call.

**Comment by kichalla** [comment]
@ycrumeyrolle Can you update your PR after addressing the comments so that I can merge this in for our next release?

**Comment by ycrumeyrolle** [comment]
@kichalla ILoggerProvider inherit IDisposable & catch exceptions in the Dispose(). Is that all?

**Comment by kichalla** [comment]
@ycrumeyrolle yes and also please rebase your PR after your changes so that can I merge it

**Comment by ycrumeyrolle** [comment]
@kichalla done

**Comment by ycrumeyrolle** [comment]
Following repos need to react to the ILoggerProvider change : 

###### Performance
https://github.com/aspnet/Performance/blob/dev/test/Microsoft.AspNet.Tests.Performance.Utility/Logging/TeamcityLoggerProvider.cs
https://github.com/aspnet/Performance/blob/dev/test/Microsoft.AspNet.Tests.Performance.Utility/Logging/ArchiveLoggerProvider.cs

###### Diagnostics
https://github.com/aspnet/Diagnostics/blob/dev/src/Microsoft.AspNet.Diagnostics.Entity/DataStoreErrorLoggerProvider.cs
https://github.com/aspnet/Diagnostics/blob/dev/src/Microsoft.AspNet.Diagnostics.Elm/ElmLoggerProvider.cs
https://github.com/aspnet/Diagnostics/blob/dev/test/Microsoft.AspNet.Diagnostics.Entity.FunctionalTests/Helpers/TestLoggerProvider.cs

###### UserSecrets
https://github.com/aspnet/UserSecrets/blob/dev/src/Microsoft.Framework.SecretManager/CommandOutputProvider.cs

###### Testing
https://github.com/aspnet/Testing/blob/dev/src/Microsoft.Framework.TestHost/TestAdapter/TestHostLoggerProvider.cs

###### EntityFramework
https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Commands/Utilities/LoggerProvider.cs
https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Commands/Executor.cs
https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Commands/Program.cs



The following repos need to react to the LoggerFactory change : 

###### Caching
https://github.com/aspnet/Caching/blob/dev/src/Microsoft.Framework.Caching.SqlConfig/Program.cs

###### EntityFramework
https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Commands/MigrationTool.cs
https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Commands/DatabaseTool.cs

###### UserSecrets
https://github.com/aspnet/UserSecrets/blob/dev/src/Microsoft.Framework.SecretManager/Program.cs

###### Hosting
https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNet.Hosting/WebHostBuilder.cs

File: src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs [hunk 4]
```diff
@@ -97,5 +99,26 @@ private static bool HasDefaultSwitch(DiagnosticsTraceSource traceSource)
...
+                }
+
+                _disposed = true;
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
```
Comment (Tratcher) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: This type does not have a finalizer.

Comment (ycrumeyrolle) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: @tracher but the type is not sealed.

Comment (Tratcher) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: Then the deriver can call GC.SuppressFinalize when they override Dispose(bool).

Comment (ycrumeyrolle) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: I am not sure to understand.
This Dispose pattern is quite common, and it is even the recommended standard implementation : 
https://msdn.microsoft.com/en-us/library/fs2xkftw(v=vs.110).aspx

And it is also implemented in aspnet :
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.ViewFeatures/Controller.cs#L1424

But if you prefer, the type can be sealed as there is no virtual method.

Comment (Tratcher) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: Note that the entire msdn write-up is discussing native resources, safe handles, finalizers, etc.. We're not doing any of that, nor do we expect anyone to derive from these types and do it.  We don't need any more than:
```c#
public void Dispose()
{
  if (!disposed)
  {
    _disposed = true;
    _rootTraceListener.Flush();
    _rootTraceListener.Dispose();
  }
}
```
That said, I don't recommend marking it sealed either, there's no reason to forbid someone from deriving from it.

Comment (ycrumeyrolle) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: Agree. 
 You recommend the same on following types ?
https://github.com/aspnet/Logging/pull/162/files#diff-f30224edfa4c3cede1d289dd6c0998eaR79
https://github.com/aspnet/Logging/pull/162/files#diff-4e0e02d9c6221bf5c7017a10fb867af3R43

And also removing comments on : 
https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Framework.Logging/Logger.cs#L141

Comment (Tratcher) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: Yes.

File: src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.Framework.Logging.TraceSource/TraceSourceLoggerProvider.cs]: Remove `IDisposable` here as `ILoggerProvider` already inherits it.

File: src/Microsoft.Framework.Logging/LoggerFactory.cs (hunk 2 above)
Comment (kichalla) [inline src/Microsoft.Framework.Logging/LoggerFactory.cs]: ILoggerFactory here should inherit from IDisposable too. (FYI: I already have a PR out for disposing this factory in hosting https://github.com/aspnet/Hosting/pull/316)

**Comment by kichalla** [comment]
:watch:

**Comment by kichalla** [comment]
:shipit: 

@Tratcher do you have any additional comments regarding this?

File: src/Microsoft.Framework.Logging.NLog/NLogLoggerProvider.cs [hunk 5]
```diff
@@ -21,6 +22,17 @@ public ILogger CreateLogger(string name)
...
         }
 
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _logFactory.Flush();
+                _logFactory.Dispose();
+                _disposed = true;
```
Comment (Tratcher) [inline src/Microsoft.Framework.Logging.NLog/NLogLoggerProvider.cs]: set _disposed before doing any work for all of these. It helps prevent concurrency, error & re-execution, and recursion problems.

**Comment by Tratcher** [comment]
One comment then :shipit:

**Comment by davidfowl** [comment]
@kichalla Whats the deal with this?

**Comment by kichalla** [comment]
commit b65fc9692043a3204662b5f4794c0930f499078e

@ycrumeyrolle Thanks for the contribution!

**Pull request closed without merging**